action/protocol/staking/protocol.go

```diff
@@ -526,13 +532,14 @@ func (p *Protocol) ReadState(ctx context.Context, sr protocol.StateReader, metho
 		resp, height, err = csr.readStateBucketCount(ctx, r.GetBucketsCount())
 	case iotexapi.ReadStakingDataMethod_CANDIDATES:
 		if epochStartHeight != 0 && p.candBucketsIndexer != nil {
-			return p.candBucketsIndexer.GetCandidates(epochStartHeight, r.GetCandidates().GetPagination().GetOffset(), r.GetCandidates().GetPagination().GetLimit())
+			resp, height, err = cssr.readStateCandidatesByIndexer(ctx, r.GetCandidates(), epochStartHeight)
```

Comment:
the code here is unclear for when and why to use `csr` and `cssr`?
Naming and adding comments might improve the readibility

Reply:
> the code here is unclear for me about when and why to use `csr` and `cssr`? Naming and adding comments might improve the readibility

sure, `csr` is for native stake state read, and the `cssr` is for stake state read including native and liquid

Reply:
> the code here is unclear for me about when and why to use `csr` and `cssr`? Naming and adding comments might improve the readibility

In `Protocol.ReadState`, all should use `cssr` instead of `csr` after support for liquid stake complete.

Reply:
Using `stakeSR.xxx` here and redirecting them to `nativeSR` via `c.nativeSR.xxx` inside `stakeSR` would express your intention.

Reply:
then `nativeSR` is constructed inside `newCompositiveStakingStateReader`